src/data/structure/UserProfile.java

```diff
@@ -57,16 +57,21 @@ public String getId() {
      * @param preferenceId
      * @return added preference
      */
-    public Preference addPreference(PreferenceCategory category, String preferenceId) {
+    public Preference addPreference(PreferenceCategory category, String preferenceId, int popularity,
```

Comment:
Completely missed the point you were making.  That is reasonable encapsulation.  Made the change.